zebra/rt_netlink.c

```diff
@@ -2243,6 +2426,84 @@ ssize_t netlink_nexthop_msg_encode(uint16_t cmd,
 				nl_attr_nest_end(&req->n, nest);
 			}
 
+			if (nh->nh_seg6local_ctx) {
+				uint32_t action;
+				uint16_t encap;
+				struct rtattr *nest;
+				const struct seg6local_context *ctx;
+
+				req->nhm.nh_family = AF_INET6;
+				action = nh->nh_seg6local_action;
+				ctx = nh->nh_seg6local_ctx;
+				encap = LWTUNNEL_ENCAP_SEG6_LOCAL;
+				nl_attr_put(&req->n, buflen, NHA_ENCAP_TYPE,
+					    &encap, sizeof(uint16_t));
+
+				nest = nl_attr_nest(&req->n, buflen,
+						    NHA_ENCAP | NLA_F_NESTED);
+				switch (action) {
+				case SEG6_LOCAL_ACTION_END:
+					nl_attr_put32(&req->n, buflen,
+						      SEG6_LOCAL_ACTION,
+						      SEG6_LOCAL_ACTION_END);
+					break;
+				case SEG6_LOCAL_ACTION_END_X:
+					nl_attr_put32(&req->n, buflen,
+						      SEG6_LOCAL_ACTION,
+						      SEG6_LOCAL_ACTION_END_X);
+					nl_attr_put(&req->n, buflen,
+						    SEG6_LOCAL_NH6, &ctx->nh6,
+						    sizeof(struct in6_addr));
+					break;
+				case SEG6_LOCAL_ACTION_END_T:
+					nl_attr_put32(&req->n, buflen,
+						      SEG6_LOCAL_ACTION,
+						      SEG6_LOCAL_ACTION_END_T);
+					nl_attr_put32(&req->n, buflen,
+						      SEG6_LOCAL_TABLE,
+						      ctx->table);
+					break;
+				case SEG6_LOCAL_ACTION_END_DX4:
+					nl_attr_put32(&req->n, buflen,
+						SEG6_LOCAL_ACTION,
+						SEG6_LOCAL_ACTION_END_DX4);
+					nl_attr_put(&req->n, buflen,
+						    SEG6_LOCAL_NH4, &ctx->nh4,
+						    sizeof(struct in_addr));
+					break;
+				case SEG6_LOCAL_ACTION_END_DT6:
+					nl_attr_put32(&req->n, buflen,
+						SEG6_LOCAL_ACTION,
+						SEG6_LOCAL_ACTION_END_DT6);
+					nl_attr_put32(&req->n, buflen,
+						      SEG6_LOCAL_TABLE,
+						      ctx->table);
+					break;
+				default:
+					zlog_err("%s: unsupport seg6local behaviour action=%u",
```

Comment:
why an error here - what could a user do with this? shouldn't we be rejecting invalid values much earlier than this? and ... if the encode fails, we should return an error instead of continuing?

Reply:
Thank you for your comment.
So we want to confirm what we need to fix here.

(1) If this function is called with an unsupported seg6local action-id, should we check for error before this switch statement and do an early return? Or (2) is it enough to do the error return in the default block of this switch statement?

(1) is like follow
```
if (action_type != TYPE_A && action_type != TYPE_B) {
  zlog_err("ERROR ...");
  return 0;
}

switch (action_type) {
  case TYPE_A:
    if (!nl_attr_put32(..)) return 0;
    if (!nl_attr_put32(..)) return 0;
  case TYPE_B:
    if (!nl_attr_put32(..)) return 0;
    if (!nl_attr_put32(..)) return 0;
}

```

(2) is like follow
```
...
switch (action_type) {
  case TYPE_A:
    if (!nl_attr_put32(..)) return 0;
    if (!nl_attr_put32(..)) return 0;
  case TYPE_B:
    if (!nl_attr_put32(..)) return 0;
    if (!nl_attr_put32(..)) return 0;
  default:
    zlog_err("ERROR ...");
    return 0;
}
...
```

I think (2) is better, because the code to check "the action type" will be redundant when we chose (1).

However, I may not have understood your comment well, so I would like to double confirm with you.

Reply:
I'll fix these lines with (2) idea.

Reply:
I fixed